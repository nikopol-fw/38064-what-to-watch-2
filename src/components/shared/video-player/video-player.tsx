import * as React from 'react';
import {RefObject} from "react";


interface Props {
  isMuted: boolean;
  poster: string;
  video: string;

  videoRef?: RefObject<HTMLVideoElement>;

  onMetadataLoaded?: () => void;
  onTimeUpdate?: () => void;
}

export const VideoPlayer: React.FC<Props> = (props) => {
  const {isMuted, poster, video, videoRef, onMetadataLoaded, onTimeUpdate} = props;

  return onMetadataLoaded && onTimeUpdate
    ? (
      <video className="player__video" preload="metadata"
        src={video}
        poster={poster}
        muted={isMuted}
        ref={videoRef}
        onLoadedMetadata={onMetadataLoaded}
        onTimeUpdate={onTimeUpdate}
      />
    ) : (
      <video className="player__video" preload="metadata"
        src={video}
        poster={poster}
        muted={isMuted}
        ref={videoRef}
      />
    );
};
