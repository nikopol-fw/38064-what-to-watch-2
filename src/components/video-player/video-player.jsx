import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';

export class VideoPlayer extends PureComponent {
  constructor(props) {
    super(props);

    this.videoRef = React.createRef();
  }

  render() {
    const {preview, poster} = this.props;

    return <video width="280" height="175" preload="metadata" muted loop
      src={preview}
      poster={`/img/${poster}`}

      ref={this.videoRef}
    />;
  }

  componentDidUpdate() {
    /** @type {HTMLVideoElement} */
    const videoElement = this.videoRef.current;
    if (this.props.isPlaying) {
      videoElement.play();
    } else {
      videoElement.load();
    }
  }
}

VideoPlayer.propTypes = {
  preview: PropTypes.string.isRequired,
  poster: PropTypes.string.isRequired,
  isPlaying: PropTypes.bool.isRequired,
};
