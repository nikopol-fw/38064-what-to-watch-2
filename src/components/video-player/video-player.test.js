import React from 'react';
import renderer from 'react-test-renderer';

import {VideoPlayer} from './video-player';


const mock = {
  isPlaying: false,
  poster: `https://htmlacademy-react-2.appspot.com/wtw/static/film/preview/gangs_of_new_york.jpg`,
  preview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
};


it(`VideoPlayer correctly renders after relaunch`, () => {
  const tree = renderer
    .create(<VideoPlayer
      poster={mock.poster}
      preview={mock.preview}
      isPlaying={mock.isPlaying}
    />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
