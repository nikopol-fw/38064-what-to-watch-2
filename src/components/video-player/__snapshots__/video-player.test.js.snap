// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPlayer correctly renders after relaunch 1`] = `
<video
  height="175"
  loop={true}
  muted={true}
  poster="https://htmlacademy-react-2.appspot.com/wtw/static/film/preview/gangs_of_new_york.jpg"
  preload="metadata"
  src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
  width="280"
/>
`;
